pkg/server/server_entrypoint_tcp.go

```diff
@@ -195,44 +210,60 @@ func (e *TCPEntryPoint) Shutdown(ctx context.Context) {
 		wg.Add(1)
 		go func() {
 			defer wg.Done()
-			if err := e.httpServer.Server.Shutdown(ctx); err != nil {
-				if ctx.Err() == context.DeadlineExceeded {
-					logger.Debugf("Wait server shutdown is overdue to: %s", err)
-					err = e.httpServer.Server.Close()
-					if err != nil {
-						logger.Error(err)
-					}
+			err := e.httpServer.Server.Shutdown(ctx)
+			if err == nil {
+				return
+			}
+			if ctx.Err() == context.DeadlineExceeded {
+				logger.Debugf("Server failed to shutdown within deadline because: %s", err)
+				if err = e.httpServer.Server.Close(); err != nil {
+					logger.Error(err)
 				}
+				return
 			}
+			logger.Error(err)
+			// We expect Close to fail again because Shutdown most likely failed when trying
+			// to close a listener. We still call it however, to make sure that all connections
+			// get closed as well.
+			e.httpServer.Server.Close()
```

Comment:
maybe we can refactor all that in a func/closure that works for both httpServer and httpsServer?